cumulus/pallets/xcmp-queue/src/migration.rs

```diff
@@ -121,95 +188,125 @@ pub mod v3 {
+				"Dropping message with format {:?} (not ConcatenatedVersionedXcm)",
+				format
+			);
+			v3::InboundXcmpMessages::<T>::remove(&next.sender, &block_number);
+			v3::InboundXcmpStatus::<T>::put(states);
+			return
+		}
 
-	T::DbWeight::get().reads_writes(overweight_messages, 1)
-}
+		let Some(msg) = v3::InboundXcmpMessages::<T>::take(&next.sender, &block_number) else {
+			defensive!("Storage corrupted: HRMP message missing:", (next.sender, block_number));
+			v3::InboundXcmpStatus::<T>::put(states);
+			return
+		};
 
-pub fn lazy_migrate_inbound_queue<T: Config>() {
-	let Some(mut states) = v3::InboundXcmpStatus::<T>::get() else {
-		log::debug!(target: LOG, "Lazy migration finished: item gone");
-		return
-	};
-	let Some(ref mut next) = states.first_mut() else {
-		log::debug!(target: LOG, "Lazy migration finished: item empty");
-		v3::InboundXcmpStatus::<T>::kill();
-		return
-	};
-	log::debug!(
-		"Migrating inbound HRMP channel with sibling {:?}, msgs left {}.",
-		next.sender,
-		next.message_metadata.len()
-	);
-	// We take the last element since the MQ is a FIFO and we want to keep the order.
-	let Some((block_number, format)) = next.message_metadata.pop() else {
-		states.remove(0);
-		v3::InboundXcmpStatus::<T>::put(states);
-		return
-	};
-	if format != XcmpMessageFormat::ConcatenatedVersionedXcm {
-		log::warn!(target: LOG,
-			"Dropping message with format {:?} (not ConcatenatedVersionedXcm)",
-			format
-		);
-		v3::InboundXcmpMessages::<T>::remove(&next.sender, &block_number);
+		let Ok(msg): Result<BoundedVec<_, _>, _> = msg.try_into() else {
+			log::error!(target: LOG, "Message dropped: too big");
+			v3::InboundXcmpStatus::<T>::put(states);
+			return
+		};
+
+		// Finally! We have a proper message.
+		T::XcmpQueue::enqueue_message(msg.as_bounded_slice(), next.sender);
+		log::debug!(target: LOG, "Migrated HRMP message to MQ: {:?}", (next.sender, block_number));
 		v3::InboundXcmpStatus::<T>::put(states);
-		return
 	}
+}
 
-	let Some(msg) = v3::InboundXcmpMessages::<T>::take(&next.sender, &block_number) else {
-		defensive!("Storage corrupted: HRMP message missing:", (next.sender, block_number));
-		v3::InboundXcmpStatus::<T>::put(states);
-		return
-	};
+pub mod v4 {
+	use super::*;
 
-	let Ok(msg): Result<BoundedVec<_, _>, _> = msg.try_into() else {
-		log::error!(target: LOG, "Message dropped: too big");
-		v3::InboundXcmpStatus::<T>::put(states);
-		return
-	};
+	/// Migrates `QueueConfigData` to v4, removing deprecated fields and bumping page
+	/// thresholds to at least the default values.
+	pub struct UncheckedMigrationToV4<T: Config>(PhantomData<T>);
+
+	impl<T: Config> OnRuntimeUpgrade for UncheckedMigrationToV4<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let translate = |pre: v2::QueueConfigData| -> QueueConfigData {
+				use sp_std::cmp::max;
+
+				let default = QueueConfigData::default();
+				let post = QueueConfigData {
+					suspend_threshold: max(pre.suspend_threshold, default.suspend_threshold),
```

Comment:
Why `max` here?

Reply:
Changed the strategy to something safer. If the values are the old defaults, migrate them to the new defaults. If not, keep them as they are. WDYT ?

Reply:
Ah okay. Since these are all thresholds they are mostly safe to be increased 👍